File: app/src/main/java/com/example/game/SpawnDirector.kt
package com.example.game

import com.example.core.Constants
import com.example.core.PowerUpType
import kotlin.random.Random

/**
 * The Spawn Director — Fairness Engine.
 * Guarantees a rolling safe corridor, readable coin arcs, kerb passengers, and fair obstacle placement.
 */
class SpawnDirector {

    private val rng = Random(System.currentTimeMillis())

    private var lastObstacleY = -100f
    private var lastPowerUpTimeMs = System.currentTimeMillis()
    private var nextPowerUpIntervalMs = (18000..26000).random().toLong()

    private var playerCurrentLane = 1
    private var playerLaneEntryTimeMs = System.currentTimeMillis()

    fun reset() {
        lastObstacleY = -100f
        lastPowerUpTimeMs = System.currentTimeMillis()
        playerCurrentLane = 1
        playerLaneEntryTimeMs = System.currentTimeMillis()
    }

    fun onPlayerLaneChanged(newLane: Int) {
        if (newLane != playerCurrentLane) {
            playerCurrentLane = newLane
            playerLaneEntryTimeMs = System.currentTimeMillis()
        }
    }

    /**
     * Safe Corridor Validator:
     * Guarantees at least 1 lane remains completely open for a valid traversable path.
     */
    fun shouldSpawnObstacleRow(
        currentY: Float,
        scrollSpeed: Float,
        tierMaxLanesBlocked: Int
    ): SpawnRowDecision? {
        val minGap = 50f + (scrollSpeed * 0.15f)
        if (currentY - lastObstacleY < minGap) {
            return null // Too close to previous row
        }

        val blockedLanes = mutableListOf<Int>()
        val lanesToBlockCount = if (tierMaxLanesBlocked >= 2 && rng.nextFloat() < 0.6f) 2 else 1

        // Rule 2: Do not spawn obstacle in the player's current lane if they entered < 0.35s ago
        val timeInLaneSec = (System.currentTimeMillis() - playerLaneEntryTimeMs) / 1000f
        val protectedLane = if (timeInLaneSec < 0.35f) playerCurrentLane else -1

        val availableLanes = (0 until Constants.LANE_COUNT).filter { it != protectedLane }.toMutableList()
        availableLanes.shuffle(rng)

        for (i in 0 until lanesToBlockCount.coerceAtMost(2)) { // Max 2 lanes blocked EVER
            if (availableLanes.isNotEmpty()) {
                blockedLanes.add(availableLanes.removeAt(0))
            }
        }

        // Rule 1: Validate at least 1 lane is ALWAYS traversable
        val openLanes = (0 until Constants.LANE_COUNT).filter { !blockedLanes.contains(it) }
        if (openLanes.isEmpty()) {
            // Unsafe! Force open at least one lane
            blockedLanes.removeAt(0)
        }

        lastObstacleY = currentY
        val obstacleType = OBSTACLE_TYPES[rng.nextInt(OBSTACLE_TYPES.size)]

        return SpawnRowDecision(blockedLanes, obstacleType)
    }

    /**
     * Generates a readable arc pattern of 5-9 Taka coins.
     */
    fun generateCoinPattern(startY: Float): List<CoinSpawnInfo> {
        val count = (5..9).random(rng)
        val coins = mutableListOf<CoinSpawnInfo>()
        val startLane = rng.nextInt(Constants.LANE_COUNT)
        val endLane = if (rng.nextBoolean()) (startLane + 1) % Constants.LANE_COUNT else (startLane + 2) % Constants.LANE_COUNT

        val kerbYGap = 16f

        for (i in 0 until count) {
            val progress = i.toFloat() / (count - 1)
            val laneFloat = startLane + (endLane - startLane) * progress
            val laneIndex = laneFloat.toInt().coerceIn(0, Constants.LANE_COUNT - 1)
            val coinX = Constants.LANE_CENTERS[laneIndex] - 4f
            val coinY = startY - (i * kerbYGap)
            coins.add(CoinSpawnInfo(coinX, coinY))
        }

        return coins
    }

    /**
     * Checks if a powerup is due to spawn (every 18-26 seconds).
     */
    fun checkPowerUpSpawn(): PowerUpType? {
        val now = System.currentTimeMillis()
        if (now - lastPowerUpTimeMs >= nextPowerUpIntervalMs) {
            lastPowerUpTimeMs = now
            nextPowerUpIntervalMs = (18000..26000).random(rng).toLong()

            // Weighted distribution: Magnet 35%, Shield 30%, Turbo 20%, Money Rain 15%
            val roll = rng.nextFloat()
            return when {
                roll < 0.35f -> PowerUpType.MONEY_MAGNET
                roll < 0.65f -> PowerUpType.SHIELD
                roll < 0.85f -> PowerUpType.TURBO
                else -> PowerUpType.MONEY_RAIN
            }
        }
        return null
    }

    data class SpawnRowDecision(val blockedLanes: List<Int>, val obstacleType: String)
    data class CoinSpawnInfo(val x: Float, val y: Float)

    companion object {
        private val OBSTACLE_TYPES = arrayOf("manhole", "garbage", "stone", "barrier", "puddle")
    }
}

File: flutter_app/lib/game/spawn_director.dart
import 'dart:math';
import '../core/constants.dart';
import '../entities/game_entities.dart';

class SpawnDirector {
  final Random rng = Random();
  double lastObstacleY = -100;
  int lastPowerUpTimeMs = DateTime.now().millisecondsSinceEpoch;
  int nextPowerUpIntervalMs = Random().nextInt(8000) + 18000;
  int playerCurrentLane = 1;
  int playerLaneEntryTimeMs = DateTime.now().millisecondsSinceEpoch;

  void reset() {
    lastObstacleY = -100;
    lastPowerUpTimeMs = DateTime.now().millisecondsSinceEpoch;
    playerCurrentLane = 1;
    playerLaneEntryTimeMs = DateTime.now().millisecondsSinceEpoch;
  }

  void onPlayerLaneChanged(int newLane) {
    if (newLane != playerCurrentLane) {
      playerCurrentLane = newLane;
      playerLaneEntryTimeMs = DateTime.now().millisecondsSinceEpoch;
    }
  }

  SpawnRowDecision? shouldSpawnObstacleRow(double currentY, double scrollSpeed, int tierMaxLanesBlocked) {
    final minGap = 50 + (scrollSpeed * 0.15);
    if (currentY - lastObstacleY < minGap) {
      return null;
    }

    final blockedLanes = <int>[];
    final lanesToBlockCount = (tierMaxLanesBlocked >= 2 && rng.nextDouble() < 0.6) ? 2 : 1;

    final timeInLaneSec = (DateTime.now().millisecondsSinceEpoch - playerLaneEntryTimeMs) / 1000.0;
    final protectedLane = timeInLaneSec < 0.35 ? playerCurrentLane : -1;

    final availableLanes = List.generate(Constants.laneCount, (i) => i).where((i) => i != protectedLane).toList();
    availableLanes.shuffle(rng);

    for (var i = 0; i < min(lanesToBlockCount, 2); i++) {
      if (availableLanes.isNotEmpty) {
        blockedLanes.add(availableLanes.removeAt(0));
      }
    }

    final openLanes = List.generate(Constants.laneCount, (i) => i).where((i) => !blockedLanes.contains(i)).toList();
    if (openLanes.isEmpty && blockedLanes.isNotEmpty) {
      blockedLanes.removeAt(0);
    }

    lastObstacleY = currentY;
    final obstacleType = obstacleTypes[rng.nextInt(obstacleTypes.length)];

    return SpawnRowDecision(blockedLanes, obstacleType);
  }

  List<CoinSpawnInfo> generateCoinPattern(double startY) {
    final count = rng.nextInt(5) + 5;
    final coins = <CoinSpawnInfo>[];
    final startLane = rng.nextInt(Constants.laneCount);
    final endLane = rng.nextBool() ? (startLane + 1) % Constants.laneCount : (startLane + 2) % Constants.laneCount;

    const kerbYGap = 16.0;

    for (var i = 0; i < count; i++) {
      final progress = i / (count - 1);
      final laneFloat = startLane + (endLane - startLane) * progress;
      final laneIndex = laneFloat.toInt().clamp(0, Constants.laneCount - 1);
      final coinX = Constants.laneCenters[laneIndex] - 4;
      final coinY = startY - (i * kerbYGap);
      coins.add(CoinSpawnInfo(coinX, coinY));
    }

    return coins;
  }

  PowerUpType? checkPowerUpSpawn() {
    final now = DateTime.now().millisecondsSinceEpoch;
    if (now - lastPowerUpTimeMs >= nextPowerUpIntervalMs) {
      lastPowerUpTimeMs = now;
      nextPowerUpIntervalMs = Random().nextInt(8000) + 18000;

      final roll = rng.nextDouble();
      if (roll < 0.35) return PowerUpType.moneyMagnet;
      if (roll < 0.65) return PowerUpType.shield;
      if (roll < 0.85) return PowerUpType.turbo;
      return PowerUpType.moneyRain;
    }
    return null;
  }
}

class SpawnRowDecision {
  final List<int> blockedLanes;
  final String obstacleType;

  SpawnRowDecision(this.blockedLanes, this.obstacleType);
}

class CoinSpawnInfo {
  final double x;
  final double y;

  CoinSpawnInfo(this.x, this.y);
}

const obstacleTypes = ["manhole", "garbage", "stone", "barrier", "puddle"];
